omniledger/service/viewchange.go

```diff
@@ -0,0 +1,371 @@
+		log.Error(s.ServerIdentity(), err)
+		return
+	}
+
+	// Store it in our log.
+	s.viewChangeMan.addReq(*req)
+}
+
+func (s *Service) startViewChangeCosi(req viewchange.NewViewReq) ([]byte, error) {
+	defer log.Lvl2(s.ServerIdentity(), "finished view-change ftcosi")
+	sb := s.db().GetByID(req.GetView().ID)
+	newRoster := rotateRoster(sb.Roster, req.GetView().LeaderIndex)
+	if !newRoster.List[0].Equal(s.ServerIdentity()) {
+		panic("startViewChangeCosi should not be called by non-leader")
+	}
+	proto, err := s.CreateProtocol(viewChangeFtCosi, newRoster.GenerateBinaryTree())
+	if err != nil {
+		panic(err)
+	}
+	payload, err := protobuf.Encode(&req)
+	if err != nil {
+		panic(err)
+	}
+
+	interval, err := s.LoadBlockInterval(req.GetView().ID)
+	if err != nil {
+		panic(err)
+	}
+
+	n := len(sb.Roster.List)
+	cosiProto := proto.(*cosiprotocol.FtCosi)
+	cosiProto.Msg = req.Hash()
+	cosiProto.Data = payload
+	cosiProto.CreateProtocol = s.CreateProtocol
+	cosiProto.Timeout = interval / 2
+	cosiProto.Threshold = n - n/3
+	cosiProto.NSubtrees = int(math.Pow(float64(n), 1.0/3.0))
+	if err := cosiProto.Start(); err != nil {
+		return nil, err
+	}
+	// The protocol should always send FinalSignature because it has a
+	// timeout, so we don't need a select.
+	return <-cosiProto.FinalSignature, nil
+}
+
+// verifyViewChange is registered in the view-change ftcosi.
+func (s *Service) verifyViewChange(msg []byte, data []byte) bool {
+	// Parse message and check hash.
+	var req viewchange.NewViewReq
+	if err := protobuf.DecodeWithConstructors(data, &req, network.DefaultConstructors(cothority.Suite)); err != nil {
+		log.Error(s.ServerIdentity(), err)
+		return false
+	}
+	if !bytes.Equal(msg, req.Hash()) {
+		log.Error(s.ServerIdentity(), "digest doesn't verify")
+		return false
+	}
+	// Check that we know about the view and the new roster in the request
+	// matches the view-change proofs.
+	sb := s.db().GetByID(req.GetView().ID)
+	if sb == nil {
+		log.Error(s.ServerIdentity(), "view does not exist")
+		return false
+	}
+	newRosterID := rotateRoster(sb.Roster, req.GetView().LeaderIndex).ID
+	if !newRosterID.Equal(req.Roster.ID) {
+		log.Error(s.ServerIdentity(), "invalid roster in request")
+		return false
+	}
+	// Check the signers are unique, they are in the roster and the
+	// signatures are correct.
+	uniqueSigners, uniqueViews := func() (int, int) {
+		signers := make(map[[16]byte]bool)
+		views := make(map[string]bool)
+		for _, p := range req.Proof {
+			signers[p.SignerID] = true
+			views[string(p.View.Hash())] = true
+		}
+		return len(signers), len(views)
+	}()
+	f := len(sb.Roster.List) / 3
+	if uniqueSigners < 2*f+1 {
+		log.Error(s.ServerIdentity(), "not enough proofs")
+		return false
+	}
+	if uniqueViews != 1 {
+		log.Error(s.ServerIdentity(), "conflicting views")
+		return false
+	}
+	for _, p := range req.Proof {
+		_, sid := sb.Roster.Search(p.SignerID)
```

Comment:
Worst case, this is O(n^2) because there's one p per server in the roster, and we search it each time. Is there a way we can use the pre-work of calculating the signers map in order to change this to one walk through sb.Roster.List, matching up Proofs to public keys in the roster?

Reply:
I'm making a map of the roster, then for every iteration in the proof we can look up an entry in the map.